### DISCUSSION TITLE
Improving Typescript code parsing in Astro components

### DISCUSSION THREAD

**Discussion by MoustaphaDev** [withastro/roadmap#819] [Proposal]
### Body

# Summary

Explore a more effective approach to parsing frontmatter code in Astro components, particularly when dealing with TypeScript.
# Background & Motivation

Currently, the Astro compiler handles TypeScript code (for export/import hoisting and `Props` type detection purposes, etc) using a JavaScript parser that it extends to support TypeScript. While functional in most cases, it may fall short when dealing with certain TypeScript code, as it's not a fully-fledged TypeScript parser. This limitation is evident in several open issues: https://github.com/withastro/compiler/issues/927, https://github.com/withastro/compiler/issues/554, https://github.com/withastro/compiler/issues/863, https://github.com/withastro/compiler/issues/763, https://github.com/withastro/compiler/issues/878, https://github.com/withastro/compiler/issues/749, https://github.com/withastro/compiler/issues/1056

While patching individual issues is possible, it's unsustainable in the long term, requiring constant maintenance and potentially introducing further incompatibilities.
# Goals

- Identify a more robust method for parsing frontmatter code
- Alleviate the maintenance burden associated with extending the current setup to fully handle TypeScript syntax

# Unresolved question

- Which TypeScript parser should be used?
- How might the transition impact performance?

**Comment by Princesseuh** [comment]
That sounds great to me! Not sure if there's many TypeScript parsers in Go, maybe ESBuild's can be used, somehow?

**Reply by natemoo-re** [reply]
We could certainly investigate forking the ESBuild TypeScript parser, like we did for ESBuild's CSS parser for similar reasons.

Unfortunately the Go ecosystem does not seem to have any mature packages that can parse modern JavaScript/TypeScript. Most are stuck on basic ES2015.

**Reply by MoustaphaDev** [reply]
Yep we'll probably have to go that route unfortunately, I'm not aware of any proper TS parser in go.

**Reply by MoustaphaDev** [reply]
Been experimenting with using https://github.com/microsoft/typescript-go this week and it's working pretty well!
It's a bummer that it doesn't have a public API though, we'd have to maintain a fork of it, but we already do vendor a few packages so yeah.

**Comment by florian-lefebvre** [comment]
I think this is now solved with the rust compiler
